fedbiomed/node/secagg.py

```diff
@@ -375,49 +381,67 @@ def _setup_specific(self) -> None:
 
-        # Key exchange with other nodes
-        # TODO: replace `dummy` with real payload
+        local_keypair = DHKey()
+        local_public_key = local_keypair.export_public_key()
+        key_agreement = DHKeyAgreement(
+            node_u_id=environ['NODE_ID'],
+            node_u_private_key=local_keypair.export_private_key(),
+            session_salt=bytes(self._secagg_id, 'utf-8'),
+        )
+
+        # Make public key available for requests received from other nodes for this `secagg_id`
+        self._controller_data.event(self._secagg_id, {'public_key': local_public_key})
 
+        # Request public key from other nodes
         other_nodes_messages = []
         for node in other_nodes:
             other_nodes_messages += [
                 NodeToNodeMessages.format_outgoing_message({
                     'request_id': REQUEST_PREFIX + str(uuid.uuid4()),
                     'node_id': environ['NODE_ID'],
                     'dest_node_id': node,
-                    'dummy': f"KEY REQUEST INNER from {environ['NODE_ID']}",
                     'secagg_id': self._secagg_id,
                     'command': 'key-request'
                 })
             ]
 
         logger.debug(f'Sending Diffie-Hellman setup for {self._secagg_id} to nodes: {other_nodes}')
-        listener_id = send_nodes(
+        all_received, messages = send_nodes(
             self._grpc_client,
             self._pending_requests,
             self._researcher_id,
             other_nodes,
             other_nodes_messages,
         )
-        all_received, messages = self._pending_requests.wait(listener_id, TIMEOUT_NODE_TO_NODE_REQUEST)
+
+        # Nota: don't clean `self._controller_data.remove(secagg_id)` when finished.
+        # Rely on automatic cleaning after timeout.
+        # This node received all replies, but some nodes may still be querying this node.
 
         logger.debug(f"Completed Diffie-Hellmann setup with success={all_received} "
                      f"node_id='{environ['NODE_ID']}' secagg_id='{self._secagg_id}")
         if not all_received:
-            nodes_no_answer = set(other_nodes) - set([m['node_id'] for m in messages])
+            nodes_no_answer = set(other_nodes) - set([m.get_param('node_id') for m in messages])
             raise FedbiomedSecaggError(
                 f"{ErrorNumbers.FB318.value}: Some nodes did not answer during Diffie Hellman secagg "
                 f"context setup: {nodes_no_answer}"
             )
 
-        # Successful DH exchange with other ndoes
-        context = { 'dummy': "tempo value to replace by LOM specific value"}
+        # At this point: successful DH exchange with other nodes
+        context = {
+            m.get_param('node_id'): key_agreement.agree(m.get_param('node_id'), m.get_param('public_key'))
+            for m in messages
+        }
+        logger.debug(f"************************ CONTEXT {context}")
```

Comment:
I don't know if do you want to keep it or remove, or remove in the final PR.

Reply:
(not existing in last committed version)